src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -208,8 +208,14 @@ def extract_trusted_facts(request_data, facts)
       end
 
       def get_facts_from_terminus(nodename, environment)
-        facts = Puppet::Node::Facts.indirection.find(nodename,
-                                                     {environment: environment})
+        # Bypass the cache terminus, which is not guaranteed to store trusted facts
+        terminus = Puppet::Node::Facts.indirection.terminus
+        request = Puppet::Indirector::Request.new(terminus.class.name,
+                                                  :find,
+                                                  nodename,
+                                                  nil,
+                                                  environment: environment)
+        facts = terminus.find(request)
```

Comment:
I think there’s an option to find to ignore the cache so you don’t need to construct the request

Reply:
Oh okay I wondered about that but wasn't sure. I can look for it.

Reply:
https://github.com/puppetlabs/puppet/blob/ba1e8af0717c6116f77d8665db115887d30c2ff1/lib/puppet/indirector/indirection.rb#L265

Reply:
Updated to use that, tested that it still works 👍